### PR TITLE
RFC: Shardonnay Partitioner

### PR DESCRIPTION
**Pull request by andydavis1**
This RFC introduces for early feedback the new Shardonnay partitioner for OpenXLA.

### CODE REVIEW

**Comment by joker-eph** [comment]
@andydavis1 (I can't find you on LLVM Discourse, do you have an account there?).

There is work going on similar mesh modeling upstream, this related work is very interesting and relevant, would you be available to present this at an MLIR Open Meeting? (That does not commit you to anything beyond that, it just seems it can be very instructive for the MLIR community at large and that can also influence the work upstream (and vice-versa maybe).

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+*   <code>partir.UNKNOWN</code>
+
+If a dimension is open we add a <code>?</code> following the axes that the dimension is already sharded on (see example below).
+
+
+#### Closed
+
+A closed dimension is one that isn’t available for propagation to add further sharding to, i.e. the specified dimension sharding is the final sharding of that dimension and it can’t be changed. A common use case of this is to make all inputs/outputs of a module static, i.e. they can't be modified.
+
+We can extend the example from above to have an open dimension and a closed dimension.
```
Comment (nouiz) [inline rfcs/2024-03-14-shardonnay-partitioner-rfc.md:131]: Would be great to show it based on the example above.

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+
+### Explicitly replicated axes
+
+An explicit set of axes that a tensor is replicated on. While it can be determined that a tensor not sharded on an axis is implicitly replicated on it, having it explicit makes sure that propagation cannot use these axes to further shard an open dimension with those axes. With implicit replication, a tensor _can_ be further partitioned. But with explicit replication, nothing can partition the tensor along that axis.
+
+Ordering of replicated axes has no effect on how the data of a tensor is stored. But, for consistency only, the axes will be stored in the order they are specified in the top level mesh. For example, if the mesh is:
```
Comment (nouiz) [inline rfcs/2024-03-14-shardonnay-partitioner-rfc.md:138]: The last sentence isn't finished.

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+And we want axes `"a"` and `"c"` to be explicitly replicated, the order should be:
```
Comment (nouiz) [inline rfcs/2024-03-14-shardonnay-partitioner-rfc.md:140]: The last sentence isn't finished.

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+We can extend our example from above to have an explicitly replicated axis.
```
Comment (nouiz) [inline rfcs/2024-03-14-shardonnay-partitioner-rfc.md:142]: Missing example?

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+However, the number of other sub-axes doesn’t make a difference when using a specific sub-axis `"x":(m)k`, and any other sub-axes don't need to be referenced in the tensor sharding if they don't shard a dimension or are explicitly replicated.
+
+Going back to the example in motivation, we can shard the result as follows:
+
+```c++
+@mesh_xy= <"x"=4>
+
+%arg0 : tensor<8xf32> {sdy.sharding=<@mesh_xy, [{"x"}]>}
+%0 = reshape %arg0 {sdy.sharding_per_value=<[<@mesh_xy, [{"x":(1)2}, {"x":(2)4}]>]>}
```
Comment (nouiz) [inline rfcs/2024-03-14-shardonnay-partitioner-rfc.md:196]: Line 185 tell: `'k>1' is the **_actual size_** of this sub-axis`
So I would have expected `(2)4` to be `(2)2`.
Any idea what I'm missing here?

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
+For example, the sharding representation in some frameworks can be an ordered list of devices and dimension sizes, but this can't be represented with the [axis splitting](?tab=t.0#heading=h.a7ox3x4zr5br) above.
+
+We propose to overcome this limitation and handle existing corner cases by defining **multiple logical meshes** at the top level of the program. Each mesh can have a different number of axes with different names, as well as its own arbitrary assignment for the same set of devices, i.e. each mesh refers to the same set of devices (by their unique logical ID) but with an arbitrary order, similar to the GSPMD representation.
+
+Each sharding representation is linked to a specific logical mesh, therefore it will only reference axes from that mesh.
+
+A tensor that is assigned to one logical mesh can be used by an op that is assigned to a different mesh, by naively resharding the tensor to match the destination mesh. In GSPMD this is what is usually done to resolve conflicting meshes.
+
+We provide two examples below:
```
Comment (nouiz) [inline rfcs/2024-03-14-shardonnay-partitioner-rfc.md:261]: Missing example?

**Review by nouiz — COMMENTED** [review]
Thanks for this interesting document.
I made a few comments.

**Comment by andydavis1** [comment]
> @andydavis1 (I can't find you on LLVM Discourse, do you have an account there?).
> 
> There is work going on similar mesh modeling upstream, this related work is very interesting and relevant, would you be available to present this at an MLIR Open Meeting? (That does not commit you to anything beyond that, it just seems it can be very instructive for the MLIR community at large and that can also influence the work upstream (and vice-versa maybe).

Thanks Mehdi. I'm andydavis1 on LLVM discourse. We are planning to be at the next openxla community meeting, but could do an MLIR Open Meeting down the road if needed...

**Comment by joker-eph** [comment]
@andydavis1 : thanks! That would be great, whenever you’re ready to chat.
Here is the upstream thread FYI: https://discourse.llvm.org/t/rfc-sharding-framework-design-for-device-mesh/73533/93

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md (hunk 1 above)
Comment (tomnatan30) [reply rfcs/2024-03-14-shardonnay-partitioner-rfc.md:131]: Done.

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md (hunk 2 above)
Comment (tomnatan30) [reply rfcs/2024-03-14-shardonnay-partitioner-rfc.md:147]: Done.

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md (hunk 3 above)
Comment (tomnatan30) [reply rfcs/2024-03-14-shardonnay-partitioner-rfc.md:153]: Done.

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md (hunk 4 above)
Comment (tomnatan30) [reply rfcs/2024-03-14-shardonnay-partitioner-rfc.md:159]: yes :)

File: rfcs/2024-03-14-shardonnay-partitioner-rfc.md (hunk 5 above)
Comment (tomnatan30) [reply rfcs/2024-03-14-shardonnay-partitioner-rfc.md]: You are absolutely right, changed.

**Comment by jpienaar** [comment]
I was OOO during the recent community meeting. Was there any questions/comments/clarification from the recent community meeting that should be added here?

**Comment by tomnatan30** [comment]
> I was OOO during the recent community meeting. Was there any questions/comments/clarification from the recent community meeting that should be added here?

I think the two two main questions were:
1. What are the advantages of GSPMD/PartIR? Which is explained briefly at the top of this RFC
2. Timelines for turning it on by default. For which we don't have a clear answer yet

**Review by jpienaar — APPROVED** [review]

**Pull request merged**